meta/store.go

```diff
@@ -994,11 +1008,28 @@ func (s *Store) Authenticate(username, password string) (ui *UserInfo, err error
 			return ErrUserNotFound
 		}
 
+		// Check the local auth cache first.
+		s.mu.RLock()
+		p, ok := s.authCache[username]
+		s.mu.RUnlock()
```

Comment:
If you go the latter path, I think you have to redo the checks under write lock, kinda like expvar does:

https://golang.org/src/expvar/expvar.go?s=2966:3008#L144